Packages/com.unity.inputsystem/InputSystem/Devices/InputDevice.cs

```diff
@@ -446,10 +446,53 @@ internal bool hasControlsWithDefaultState
         internal void SetUsage(InternedString usage)
         {
             // Make last entry in m_UsagesForEachControl be our device usage string.
+            SetUsage(m_UsageToControl != null ? m_UsageToControl.Length : 0, usage);
+        }
+
+        internal void AddUsage(InternedString usage)
+        {
+            // Check if the usage already exists
+            if (Array.Exists(m_UsagesForEachControl, o => o == usage))
+                return;
+
+            SetUsage(m_UsagesForEachControl != null ? m_UsagesForEachControl.Length : 0, usage);
+        }
+
+        internal void RemoveUsage(InternedString usage)
+        {
             var numControlUsages = m_UsageToControl != null ? m_UsageToControl.Length : 0;
-            Array.Resize(ref m_UsagesForEachControl, numControlUsages + 1);
-            m_UsagesForEachControl[numControlUsages] = usage;
-            m_UsagesReadOnly = new ReadOnlyArray<InternedString>(m_UsagesForEachControl, numControlUsages, 1);
+            var usageIndex = Array.IndexOf(m_UsagesForEachControl, usage, numControlUsages);
+            if (usageIndex < numControlUsages)
+                return;
+
+            var usageList = new System.Collections.ArrayList(m_UsagesForEachControl);
+            var commonPart = Array.FindAll<InternedString>((InternedString[])usageList.GetRange(numControlUsages, usageList.Count - numControlUsages).ToArray(typeof(InternedString)), e => e != usage);
+
+            Array.Resize(ref m_UsagesForEachControl, numControlUsages); 
+            ArrayHelpers.Append(ref m_UsagesForEachControl, commonPart);
```

Comment:
Use `ArrayHelpers.EraseAt` instead of these four lines. The current code needless allocate several temporary objects.

```
ArrayHelpers.EraseAt(ref m_UsagesForEachControl, usageIndex);
```